src/evict.c

```diff
@@ -592,12 +598,18 @@ int performEvictions(void) {
                  * every DB. */
                 for (i = 0; i < server.dbnum; i++) {
                     db = server.db+i;
-                    dict = (server.maxmemory_policy & MAXMEMORY_FLAG_ALLKEYS) ?
-                           getRandomDict(db) : db->expires;
-                    if ((keys = dictSize(dict)) != 0) {
-                        evictionPoolPopulate(i, dict, db, pool);
-                        total_keys += keys;
-                    }
+                    do {
+                        dict = (server.maxmemory_policy & MAXMEMORY_FLAG_ALLKEYS) ?
+                               getRandomDict(db) : db->expires;
+                        if ((keys = dictSize(dict)) != 0) {
+                            evictionPoolPopulate(i, dict, db, pool);
+                            total_keys += keys;
+                        }
```

Comment:
In the average case, when keys are randomly distributed across slots, it shouldn't matter if we pick multiple keys from one slot or different slots for sampling. The fact that keys have same CRC hash should not matter for LRU algorithm efficiency. I think current implementation makes a good tradeoff between general case performance and edge case efficiency.

Reply:
well, truth be told, even before that, dictGetSomeKeys would keep getting more and more keys from the same dict bucket, and then move to the next one (without calling random again).
but in case it hits a stride of empty buckets, it'll then re-random for a different bucket.

so first, maybe with the new approach, we should have it exit and re-random for another slot?

secondly, to my original point, you're right that for LRU related efficiency it doesn't matter much, but it could cause some full slots to be emptied (finding the best LRU just from within these slots), and then after eviction, keys will not be evenly distributed across the slots. i suppose this is mainly a concern on small databases, with not many keys per slot.
maybe considering that we take only 5 keys from each slot and then re-random, it's not that bad on a common case.

Reply:
Please look at the most recent version where fairness issues were addressed. This comment explains approach that I took https://github.com/redis/redis/pull/11695#issuecomment-1480454348